### ISSUE TITLE
refactor(ast): add a container macro to AST nodes

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#2641]
@overlookmotel originally said on #2640 (splitting into separate issue for clarity):

We could add an `Arena` trait to all AST types to prevent allocation-owning types getting into the arena and resulting in memory leaks.

The most ergonomic implementation would be a terse attribute macro `#[arena]`.

We could also add to that macro a bunch of the boilerplate attrs which are crowding `oxc_ast`. It could also allow `#[serde]` attrs to be used without `#[cfg_attr]`.

I know that macros can obscure meaning, and explicit code with no magic is a virtue, but personally I find the level of attributes in `oxc_ast` is now so high that it makes it hard to see the wood for the trees. I think this:

```rs
#[arena]
#[serde(tag = "type")]
pub struct Program<'a> {
  #[serde(flatten)]
  pub span: Span,
  // ...
}
```

is much easier to grok than:

```rs
#[derive(Debug, Hash)]
#[cfg_attr(feature = "serde", derive(Serialize), serde(tag = "type", rename_all = "camelCase"))]
#[cfg_attr(all(feature = "serde", feature = "wasm"), derive(tsify::Tsify))]
pub struct Program<'a> {
  #[cfg_attr(feature = "serde", serde(flatten))]
  pub span: Span,
  // ...
}
```

If the macro's implementation is simple enough, and we added a comment to the top of each file saying `// See 'arena' macro in 'oxc_macros' crate for attributes which are added to AST types`, then that'd make it easier for people new to the codebase to understand it swiftly.

I do have one further motivation: A macro of this sort would also be very useful for AST transfer, which needs to mutate enums extensively (see current impl [here](https://github.com/oxc-project/oxc/pull/2457/files#diff-ae923e4dbf27e27ba280646743eb2aedaf8e718fe32b5e5a0da2f88ba5165950)).

Of course the big argument against an `#[arena]` macro is if it'd have a noticeable impact on compile times. If it's going to hurt that measure considerably, then probably a bad idea. Do you think that'd be the case?

**Comment by Boshen** [comment]
### Shrinking derive macros

Readability wise it's 💩, but everything is googleable if it's all laid out. New contributors will have a hard time navigating things if everything is hidden away.

Sorry I'm in the "less / no macros" camp 😅

**Comment by overlookmotel** [comment]
I do agree with you in general about macros, but in the case of `oxc_ast`, I just think it's getting out of hand with the sheer volume of `#[cfg_attr]`s everywhere. It's becoming seriously hard to see what the hell is going on - and I say that as someone who's been heavily immersed in OXC's codebase for 2-3 months now.

These `#[cfg_attr]`s are *almost* the same for all the AST types. Worse, there are *some* differences - but those differences are wrapped in so much boilerplate, that it's hard to see them.

In my opinion, the reason macros make codebases hard to follow is not that macros are a bad thing in themselves, but lazy development practices:

1. Lack of documentation - `#[macro]` with no comment in the file *where it's used* saying why it's there, what it does, or where to find the code for the macro.
2. Macros too "clever" - convoluted macro code which makes it completely impossible to understand what a macro expands to, and no comments to explain it.

OXC's development practices are *not* lazy, and I think we can avoid these pitfalls.

So, yes, there are serious problems with how macros are often used, and in general it's best to avoid them. But I do think a pragmatic approach which recognises that *sometimes* they're appropriate is helpful. And in my view this is one such case.

I'm arguing fairly hard on this, but that's just my personality! Please do feel free to disagree.

**Comment by overlookmotel** [comment]
How about this for a compromise?

```rs
/// Dummy derive macro which doesn't actually derive anything,
/// but allows `#[serde]` attrs on the item, without deriving `Serialize`
/// and without `#[cfg_attr]`.
#[proc_macro_derive(SerdeSupport, attributes(serde))]
pub fn serde_support(_input: TokenStream) -> TokenStream {
    // Return an empty token stream
    TokenStream::new()
}
```

Then this:

```rs
#[derive(Debug, Hash)]
#[cfg_attr(feature = "serde", derive(Serialize), serde(tag = "type", rename_all = "camelCase"))]
#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
pub struct Program<'a> {
    #[cfg_attr(feature = "serde", serde(flatten))]
    pub span: Span,
    // ...
}
```

becomes this:

```rs
#[derive(Debug, Hash, SerdeSupport)]
#[cfg_attr(feature = "serde", derive(Serialize))]
#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
#[serde(tag = "type", rename_all = "camelCase")]
pub struct Program<'a> {
    #[serde(flatten)]
    pub span: Span,
    // ...
}
```

Nothing is hidden away in a macro, but the `#[cfg_attr]` boilerplate is separated from the `#[serde]` tags which contain the meaningful information.

**Comment by Boshen** [comment]
👍 go for it
